### PR TITLE
Replay buffer

### PR DESCRIPTION
**Pull request by theovincent**
Here is a first version of replay buffer. It is implemented with a BaseReplayBuffer.

There is also a refactoring of the agents with a BaseAgent.

This is an example of how to interact with the replay buffer, the values `FAKE_VALUE` and `FAKE_LOG_PROB` with have to be computed in the *real* loop:
```Python
from acme import specs

import dm_helper as helpers
from agents import RandomAgent

from replay_buffers import FixedReplayBuffer

FAKE_VALUE = 1.
FAKE_LOG_PROB = 0.4

environment = helpers.make_environment("Pendulum-v1")
environment_spec = specs.make_environment_spec(environment)
agent = RandomAgent(environment_spec)

replay_buffer = FixedReplayBuffer(10)

replay_buffer.add_first(environment.reset())
episode_return = 0
steps = 0

while not replay_buffer.timestep.last() and steps < 3:
    action = agent.select_action(replay_buffer.timestep.observation)
    timestep = environment.step(action)
    replay_buffer.add(FAKE_VALUE, FAKE_LOG_PROB, action, timestep)

    steps += 1
    episode_return += timestep.reward

batch_transitions = replay_buffer.sample_full_buffer()
batch_transitions.observation_t
``` 

Here is an example to show that the agent is hidding the replay buffer:
```Python
import jax
import chex
from acme import specs

import dm_helper as helpers
from networks import PolicyNetwork, ValueNetwork
from agents import VanillaPPO

environment = helpers.make_environment("Pendulum-v1")
environment_spec = specs.make_environment_spec(environment)

FAKE_VALUE = 1.
FAKE_LOG_PROB = 0.4

def policy_network(observations: chex.Array):
    return PolicyNetwork((10, 20), environment_spec.actions, "policy")(observations)
def value_network(observations: chex.Array):
    return ValueNetwork((10, 20), "value")(observations)
key = jax.random.PRNGKey(1)

agent = VanillaPPO(
        environment_spec.observations,
        policy_network,
        value_network,
        key,
        learning_rate=1e-3,
        discount=0.99,
    )


timestep = environment.reset()
agent.observe_first(timestep)
episode_return = 0
steps = 0

while not timestep.last() and steps < 3:
    action = agent.select_action(timestep.observation)
    timestep = environment.step(action)
    agent.observe(FAKE_VALUE, FAKE_LOG_PROB, action, timestep)

    steps += 1
    episode_return += timestep.reward

batch_transitions = agent.replay_buffer.sample_full_buffer()
batch_transitions.observation_t
```
In both scripts you should expect something like:
```
DeviceArray([[-0.46210587, -0.8868248 , -0.58006275],
             [-0.46210587, -0.8868248 , -0.58006275],
             [-0.46210587, -0.8868248 , -0.58006275]], dtype=float32)
```

### CODE REVIEW

File: ppo/replay_buffers/base_replay_buffer.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        """Randomly sample a transition from memory."""
+        assert len(self._memory) > 0, "replay buffer is unfilled"
+
+        transition_idx = np.random.randint(0, len(self._memory))
+        transition = self._memory.pop(transition_idx)
+
+        return transition
+
+    def sample_full_buffer(self) -> chex.Array:
+        """Randomly sample a transition from memory."""
```
Comment (nviolante25) [inline ppo/replay_buffers/base_replay_buffer.py:26]: Maybe we should change this docstring; it's the same as for the `sample_a_transition` method but they do different things.

File: ppo/replay_buffers/base_replay_buffer.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    def __init__(self, buffer_capacity: int) -> None:
+        self._memory = list()
+        self._maxlen = buffer_capacity
+
+    def sample_a_transition(self) -> Transition:
+        """Randomly sample a transition from memory."""
+        assert len(self._memory) > 0, "replay buffer is unfilled"
+
+        transition_idx = np.random.randint(0, len(self._memory))
```
Comment (nviolante25) [inline ppo/replay_buffers/base_replay_buffer.py:20]: Do you think we should use jax sampling with a `rng` here? Using numpy it's fine for me, just I was just wondering.

Comment (nviolante25) [inline ppo/replay_buffers/base_replay_buffer.py:20]: Do you think we should use jax sampling with a `rng` here? Using numpy it's fine for me, just I was just wondering.

File: ppo/replay_buffers/base_replay_buffer.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        transition = self._memory.pop(transition_idx)
```
Comment (nviolante25) [inline ppo/replay_buffers/base_replay_buffer.py:21]: For training ppo we are going to use the full dataset in the buffer for several epochs. This `pop` method seems to delete the dataset after the first epoch.

Comment (nviolante25) [inline ppo/replay_buffers/base_replay_buffer.py:21]: For training ppo we are going to use the full dataset in the buffer for several epochs. This `pop` method seems to delete the dataset after the first epoch.

File: ppo/agents/base_agent.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        self.value_params = self.value_network.init(
+            rng=value_key, observations=jnp.zeros(observation_spec.shape)
+        )
+
+        self.policy_optimizer = optax.adam(learning_rate)
+        self.policy_optimizer_state = self.policy_optimizer.init(self.policy_params)
+        self.value_optimizer = optax.adam(learning_rate)
+        self.value_optimizer_state = self.value_optimizer.init(self.value_params)
+
+        def greedy_policy(
```
Comment (emasquil) [inline ppo/agents/base_agent.py:39]: maybe change its name to `sampling_policy`

File: ppo/replay_buffers/base_replay_buffer.py (hunk 2 above)
Comment (theovincent) [reply ppo/replay_buffers/base_replay_buffer.py:20]: I think it would be nice yes, let's do it!

File: ppo/replay_buffers/base_replay_buffer.py (hunk 3 above)
Comment (theovincent) [reply ppo/replay_buffers/base_replay_buffer.py:21]: I will do a dataloader then :)

**Review by emasquil — COMMENTED** [review]
This is really good! We'll just wait to have the final version on the sampler of the buffer before merging it

**Comment by theovincent** [comment]
This is the new version of the branch :0!
Here is the code that shows how to interact with the dataloader:
```Python
import jax
import chex
from acme import specs

import dm_helper as helpers
from networks import PolicyNetwork, ValueNetwork
from agents import VanillaPPO
from replay_buffers import DataLoader

environment = helpers.make_environment("Pendulum-v1")
environment_spec = specs.make_environment_spec(environment)

FAKE_VALUE = 1.
FAKE_LOG_PROB = 0.4
batch_size = 2

def policy_network(observations: chex.Array):
    return PolicyNetwork((10, 20), environment_spec.actions, "policy")(observations)
def value_network(observations: chex.Array):
    return ValueNetwork((10, 20), "value")(observations)
key = jax.random.PRNGKey(1)

agent = VanillaPPO(
        environment_spec.observations,
        policy_network,
        value_network,
        key,
        learning_rate=1e-3,
        discount=0.99,
    )

timestep = environment.reset()
agent.observe_first(timestep)
episode_return = 0
steps = 0

while not timestep.last() and steps < 7:
    action = agent.select_action(timestep.observation)
    timestep = environment.step(action)
    agent.observe(FAKE_VALUE + steps, FAKE_LOG_PROB, action, timestep)

    steps += 1
    episode_return += timestep.reward

dataloarder = DataLoader(agent.replay_buffer, batch_size)
dataloarder.shuffle()

for transition in dataloarder:
    print(transition.value_t)
    print()
```
Even if it is "random", you should get the exact same output as:
```
[2. 5.]

[4. 1.]

[3. 7.]

[6.]
```
`dataloarder.get_full_memory()` gives you the trajectory in the original order even if the dataloader has already been suffled.
For now we don't use the methods `sample_pop_a_transition` and `sample_empty_full_buffer`, I kept them in case we need them in the future...

**Comment by paulinesert** [comment]
Great for me :) 
really no big deal it's written "suffling_keys" i suppose you meant "shuffling_keys" ahah :) 
But otherwise it's great ! Thank you for that 
Good to merge for me

**Review by emasquil — APPROVED** [review]

**Review by nviolante25 — APPROVED** [review]
